src/storage/index.rs

```
@@ -90,53 +87,135 @@ impl Block {
 			return Err(anyhow::anyhow!("header has overwritten data"));
 		}
 
-		self.header = Some(header);
+		return Result::Ok(BlockFile { header, data: self });
+	}
+}
+
+#[derive(Debug)]
+pub struct BlockFile {
+	header: BlockHeader,
+	data: BlockData,
+}
 
-		return Result::Ok(());
+fn header_size(size: usize) -> u64 {
+	// (struct byte) +
+	// (3 * u64 = tags + keys + timestamps) +
+	// (max array overhead) +
+	// (size * u64) +
+	// (2 * u64 = from + to)
+	return (1 + 3 * 9 + 5 + size * 9 + 2 * 9) as u64;
```

если результат выражения не влезает в `u64`, останутся только нижние биты
(в мистической ситуации, в которой `size_of::<usize>() > size_of::<u64>()`, но всё же)
мы точно не хотим паниковать в этом случае? Что-нибудь вроде
```
return (...expr...).try_into().unwrap();
```

---

ну сюда по идее передается размер массива, и если у тебя размер массива не влезает в u64 я думаю у тебя проблемы начнутся намного раньше))

---

в любом случае, проще fail fast, чем silently do the wrong thing

---

а нафига проверять то, что по сути не может произойти никак?

я просто фиксил кучу багов с памятью, гонками, логикой, но, мне кажется, ни разу не фиксил ошибку с тем, что usize в u64 не влез...